File: rps/solutions/rps_logic.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;

char MOVES[] = {'R', 'P', 'S'};

int round_num = 0;
char my_history[100000], bot_history[100000];

char play_move() {
    char move;
    if (round_num >= 2 && my_history[round_num - 1] == my_history[round_num - 2]) {
        switch (my_history[round_num - 1]) {
            case 'R':
                move = 'S';
                break;
            case 'P':
                move = 'R';
                break;
            default:
                move = 'P';
        }
    } else if (round_num >= 2 && bot_history[round_num - 1] == bot_history[round_num - 2]) {
        move = bot_history[round_num - 1];
    } else {
        move = MOVES[rand() % 3];
    }
    
    my_history[round_num] = move;
    return move;
}

void read_bot_move(char bot_move) {
    bot_history[round_num] = bot_move;
}

int main() {
    // Solutions should be deterministic, so we need to set our random seed
    srand(1337);
    
    int T;
    cin >> T;
    for (int i = 0; i < T; i++) {        
        char my_move;
        my_move = play_move();
        cout << my_move << endl;
        
        char bot_move;
        cin >> bot_move;
        read_bot_move(bot_move);
        
        round_num++;
    }
}

File: rps/solutions/rps_logic.py
import random

# Solutions should be deterministic, so we need to set our random seed
random.seed(1337)

my_history, bot_history = [], []

def play_move() -> str:
    if len(my_history) >= 2 and my_history[-1] == my_history[-2]:
        if my_history[-1] == 'R':
            move = 'S'
        elif my_history[-1] == 'P':
            move = 'R'
        else:
            move = 'P'
    elif len(bot_history) >= 2 and bot_history[-1] == bot_history[-2]:
        move = bot_history[-1]
    else:
        move = random.choice('RPS')
    
    my_history.append(move)
    return move

def read_bot_move(bot_move: str) -> None:
    bot_history.append(bot_move)

def main():
    T = int(input())
    for _ in range(T):
        print(play_move())
        read_bot_move(input())

if __name__ == '__main__':
    main()
